In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import path of data
path = r'C:\Users\nikik\17-12-2025 Instacart Basket Analysis'

In [3]:
# 1. importing ords_prods_merge dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'df_ords_prods_merge_new_columns.pkl'))

In [4]:
ords_prods_merge.shape

(32404859, 20)

# 2. Aggregating Data

In [5]:
# peforming single aggregation
department_mean_orders = ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

In [6]:
department_mean_orders

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# 3. The average order_number varies across departments, with department 21 having the highest mean orders (22.9) and department 5 among the lowest (15.2). This indicates that some departments are consistently reordered more frequently. Compared to a subset, the full dataset provides a more accurate and representative picture of ordering patterns across all departments.

# 4. Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [7]:
# Creating loyalt flag
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  ...  \
0           1987        1987   3139998      138            28  ...   
1           1989        1989   1977647      138            30  ...   
2          11433       11433    389851      709             2  ...   
3          12198       12198    652770      764             1  ...   
4          12200       12200   1813452      764             3  ...   
5          12372       12372   1701441      777            16  ...   
6          13096       13096   1871483      825             3  ...   
7          14603       14603   1290456      910            12  ...   
8          17065       17065    369558     1052            10  ...   
9          17070       17070    589712     1052            15  ...   
10         22082       22082   1798115     1379             1  ...   
11         23713       23713    497193     1480            36  ...   
12         23715       23715   1727826     1480            38  ...   
13         23716       23716   2694698     1480            39  ...   
14         23937       23937   3171753     1494             6  ...   

    order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                  11                     3.0                  5          0   
1                  17                    20.0                  1          1   
2                  21                     6.0                 20          0   
3                  13                     0.0                 10          0   
4                  17                     9.0                 11          1   
5                   7                    26.0                  7          0   
6                  14                    30.0                  2          0   
7                  10                    30.0                  1          0   
8                  20                    19.0                  1          0   
9                  12                    15.0                  2          1   
10                 11                     0.0                  1          0   
11                 13                     7.0                  1          0   
12                 13                     7.0                  3          1   
13                 12                    12.0                  4          1   
14                 10                     2.0                  8          0   

    _merge    price_range_loc     busiest_day  busiest_days  \
0     both  Mid-range product  Regularly busy  Regular days   
1     both  Mid-range product  Regularly busy  Regular days   
2     both  Mid-range product     Busiest day  Busiest days   
3     both  Mid-range product  Regularly busy  Slowest days   
4     both  Mid-range product      Lea

In [9]:
# Deriving Columns with loc()
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [10]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [11]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [13]:
# Checking the output
ords_prods_merge.head(15)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   

    Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  ...  \
0           1987        1987   3139998      138            28  ...   
1           1989        1989   1977647      138            30  ...   
2          11433       11433    389851      709             2  ...   
3          12198       12198    652770      764             1  ...   
4          12200       12200   1813452      764             3  ...   
5          12372       12372   1701441      777            16  ...   
6          13096       13096   1871483      825             3  ...   
7          14603       14603   1290456      910            12  ...   
8          17065       17065    369558     1052            10  ...   
9          17070       17070    589712     1052            15  ...   
10         22082       22082   1798115     1379             1  ...   
11         23713       23713    497193     1480            36  ...   
12         23715       23715   1727826     1480            38  ...   
13         23716       23716   2694698     1480            39  ...   
14         23937       23937   3171753     1494             6  ...   

    days_since_prior_order  add_to_cart_order  reordered  _merge  \
0                      3.0                  5          0    both   
1                     20.0                  1          1    both   
2                      6.0                 20          0    both   
3                      0.0                 10          0    both   
4                      9.0                 11          1    both   
5                     26.0                  7          0    both   
6                     30.0                  2          0    both   
7                     30.0                  1          0    both   
8                     19.0                  1          0    both   
9                     15.0                  2          1    both   
10                     0.0                  1          0    both   
11                     7.0                  1          0    both   
12                     7.0                  3          1    both   
13                    12.0                  4          1    both   
14                     2.0                  8          0    both   

      price_range_loc     busiest_day  busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regular days           Most orders   
1   Mid-range product  Regularly busy  Regular days        Average orders   
2   Mid-range product     Busiest day  Busiest days        Average orders   
3   Mid-range product  Regularly busy  Slowest days           Most orders   
4   Mid-range product      Least busy  Slowest days        Average orders   
5   Mid-range product  Regularly busy  Busiest days        Average 

# 5. The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 

In [14]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

# Analysis of Spending Habits by Customer Type
  The average product prices differ slightly across customer types: loyal customers tend to purchase slightly cheaper items on average (10.39) compared to regular (12.50) and new customers (13.29). The minimum and maximum prices are the same for all groups, suggesting that while all customers occasionally buy very cheap or very expensive products, loyal customers generally spend less per order on average.

# Create a spending flag

In [15]:
# Aggregating data
ords_prods_merge['avg_product_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [16]:
ords_prods_merge[['user_id' , 'avg_product_price']].head(15)

user_id  avg_product_price
0       138           6.935811
1       138           6.935811
2       709           7.930208
3       764           4.972414
4       764           4.972414
5       777           6.935398
6       825           5.957576
7       910           6.680000
8      1052           7.162500
9      1052           7.162500
10     1379           5.813636
11     1480           5.081633
12     1480           5.081633
13     1480           5.081633
14     1494           6.803797

# # 6.  Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

In [18]:
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] < 10, 'spending_flag'] = 'Low spender'

# If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [19]:
ords_prods_merge.loc[ords_prods_merge['avg_product_price'] >= 10, 'spending_flag'] = 'High spender'

In [20]:
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [21]:
# checking relevant column
ords_prods_merge[['user_id', 'avg_product_price', 'spending_flag']].head(15)

user_id  avg_product_price spending_flag
0       138           6.935811   Low spender
1       138           6.935811   Low spender
2       709           7.930208   Low spender
3       764           4.972414   Low spender
4       764           4.972414   Low spender
5       777           6.935398   Low spender
6       825           5.957576   Low spender
7       910           6.680000   Low spender
8      1052           7.162500   Low spender
9      1052           7.162500   Low spender
10     1379           5.813636   Low spender
11     1480           5.081633   Low spender
12     1480           5.081633   Low spender
13     1480           5.081633   Low spender
14     1494           6.803797   Low spender

# 7. In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag

In [22]:
# Aggregating data
ords_prods_merge['median_days_between_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [23]:
# checking relevant columns
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders']].head(15)

user_id  days_since_prior_order  median_days_between_orders
0       138                     3.0                         8.0
1       138                    20.0                         8.0
2       709                     6.0                         6.0
3       764                     0.0                         9.0
4       764                     9.0                         9.0
5       777                    26.0                        11.0
6       825                    30.0                        19.0
7       910                    30.0                         3.5
8      1052                    19.0                         9.5
9      1052                    15.0                         9.5
10     1379                     0.0                        20.0
11     1480                     7.0                         7.0
12     1480                     7.0                         7.0
13     1480                    12.0                         7.0
14     1494                     2.0                         6.0

# If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”

In [24]:
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

# If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”

In [25]:
ords_prods_merge.loc[(ords_prods_merge['median_days_between_orders'] <= 20) & (ords_prods_merge['median_days_between_orders'] > 10), 'order_frequency_flag'] = 'Regular customer'

# If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [26]:
ords_prods_merge.loc[ords_prods_merge['median_days_between_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [27]:
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        22796659
Regular customer          6921472
Non-frequent customer     2686728
Name: order_frequency_flag, dtype: int64

In [28]:
ords_prods_merge[['user_id', 'days_since_prior_order', 'median_days_between_orders', 'order_frequency_flag']].head(15)

user_id  days_since_prior_order  median_days_between_orders  \
0       138                     3.0                         8.0   
1       138                    20.0                         8.0   
2       709                     6.0                         6.0   
3       764                     0.0                         9.0   
4       764                     9.0                         9.0   
5       777                    26.0                        11.0   
6       825                    30.0                        19.0   
7       910                    30.0                         3.5   
8      1052                    19.0                         9.5   
9      1052                    15.0                         9.5   
10     1379                     0.0                        20.0   
11     1480                     7.0                         7.0   
12     1480                     7.0                         7.0   
13     1480                    12.0                         7.0   
14     1494                     2.0                         6.0   

   order_frequency_flag  
0     Frequent customer  
1     Frequent customer  
2     Frequent customer  
3     Frequent customer  
4     Frequent customer  
5      Regular customer  
6      Regular customer  
7     Frequent customer  
8     Frequent customer  
9     Frequent customer  
10     Regular customer  
11    Frequent customer  
12    Frequent customer  
13    Frequent customer  
14    Frequent customer

# Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [29]:
ords_prods_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_grouped.pkl'))